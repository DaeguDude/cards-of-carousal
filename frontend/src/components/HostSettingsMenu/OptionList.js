import React from 'react';
import PropTypes from 'prop-types';
import styled, { css } from 'styled-components';

const propTypes = {
  listContent: PropTypes.arrayOf(PropTypes.string).isRequired,
  isActive: PropTypes.bool.isRequired,
  onAccordionClick: PropTypes.func.isRequired,
  onListItemClick: PropTypes.func.isRequired,
  activeText: PropTypes.string.isRequired,
  inactiveText: PropTypes.string.isRequired,
};

const OptionListButton = styled.button`
  color: var(--primary-text-color);
  background-color: var(--primary-background-color);

  ${(props) =>
    props.isActive &&
    css`
      color: var(--secondary-text-color);
      background-color: var(--secondary-background-color);
    `}

  display: flex;
  flex-direction: column;
  border: none;
  padding: 20px 25px;
  width: 100%;

  text-align: left;
  font-size: 1.5rem;
  font-weight: 700;
`;

const OptionListItemButton = styled(OptionListButton)`
  font-size: 1.3rem;
  padding: 10px 50px;
`;

function OptionList({
  listContent,
  isActive: optionListIsActive,
  onAccordionClick: onOptionListClick,
  onListItemClick,
  activeText,
  inactiveText,
}) {
  return (
    <div>
      <OptionListButton
        type="button"
        onClick={onOptionListClick}
        isActive={optionListIsActive}
      >
        {optionListIsActive ? activeText : inactiveText}
      </OptionListButton>
      {optionListIsActive &&
        listContent.map((listItem) => (
          <OptionListItemButton
            type="button"
            key={listItem}
            onClick={(event) => onListItemClick(event, listItem)}
          >
            {listItem}
          </OptionListItemButton>
        ))}
    </div>
  );
}

OptionList.propTypes = propTypes;

export default OptionList;
