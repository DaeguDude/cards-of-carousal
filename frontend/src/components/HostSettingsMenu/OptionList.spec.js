import React from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import renderer from 'react-test-renderer';

import OptionList from './OptionList';

describe('OptionButton', () => {
  describe('rendering', () => {
    it('renders a button with the given open text when open', () => {
      render(
        <OptionList
          listContent={['foo', 'bar', 'bash']}
          isOpen
          onOptionListClick={() => {}}
          onListItemClick={() => {}}
          openText="LIST IS OPEN"
          closedText="LIST IS CLOSED"
        />,
      );

      expect(
        screen.getByRole('button', { name: 'LIST IS OPEN' }),
      ).toBeInTheDocument();
    });

    it('renders a button with the given closed text when closed', () => {
      render(
        <OptionList
          listContent={['foo', 'bar', 'bash']}
          isOpen={false}
          onOptionListClick={() => {}}
          onListItemClick={() => {}}
          openText="LIST IS OPEN"
          closedText="LIST IS CLOSED"
        />,
      );

      expect(
        screen.getByRole('button', { name: 'LIST IS CLOSED' }),
      ).toBeInTheDocument();
    });

    it('does not render the contents of the list when closed', () => {
      render(
        <OptionList
          listContent={['foo', 'bar', 'bash']}
          isOpen={false}
          onOptionListClick={() => {}}
          onListItemClick={() => {}}
          openText="LIST IS OPEN"
          closedText="LIST IS CLOSED"
        />,
      );

      expect(
        screen.queryByRole('button', { name: 'foo' }),
      ).not.toBeInTheDocument();
      expect(
        screen.queryByRole('button', { name: 'bar' }),
      ).not.toBeInTheDocument();
      expect(
        screen.queryByRole('button', { name: 'bash' }),
      ).not.toBeInTheDocument();
    });

    it('renders the content of the list when open', () => {
      render(
        <OptionList
          listContent={['foo', 'bar', 'bash']}
          isOpen
          onOptionListClick={() => {}}
          onListItemClick={() => {}}
          openText="LIST IS OPEN"
          closedText="LIST IS CLOSED"
        />,
      );

      expect(screen.getByRole('button', { name: 'foo' })).toBeInTheDocument();
      expect(screen.getByRole('button', { name: 'bar' })).toBeInTheDocument();
      expect(screen.getByRole('button', { name: 'bash' })).toBeInTheDocument();
    });

    it('matches the expected snapshot when open', () => {
      const tree = renderer
        .create(
          <OptionList
            listContent={['foo', 'bar', 'bash']}
            isOpen
            onOptionListClick={() => {}}
            onListItemClick={() => {}}
            openText="LIST IS OPEN"
            closedText="LIST IS CLOSED"
          />,
        )
        .toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('matches the expected snapshot when not open', () => {
      const tree = renderer
        .create(
          <OptionList
            listContent={['foo', 'bar', 'bash']}
            isOpen={false}
            onOptionListClick={() => {}}
            onListItemClick={() => {}}
            openText="LIST IS OPEN"
            closedText="LIST IS CLOSED"
          />,
        )
        .toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  describe('functionality', () => {
    it('does not call the onOptionListClick callback when the OptionList main button is not clicked', () => {
      const consoleSpy = jest
        .spyOn(console, 'error')
        .mockImplementation(() => {});

      render(
        <OptionList
          listContent={['foo', 'bar', 'bash']}
          isOpen
          onOptionListClick={consoleSpy}
          onListItemClick={() => {}}
          openText="LIST IS OPEN"
          closedText="LIST IS CLOSED"
        />,
      );

      expect(consoleSpy).not.toHaveBeenCalled();
    });

    it('calls the onOptionListClick callback when the OptionList main button is open and is clicked', () => {
      const consoleSpy = jest
        .spyOn(console, 'error')
        .mockImplementation(() => {});

      render(
        <OptionList
          listContent={['foo', 'bar', 'bash']}
          isOpen
          onOptionListClick={consoleSpy}
          onListItemClick={() => {}}
          openText="LIST IS OPEN"
          closedText="LIST IS CLOSED"
        />,
      );

      userEvent.click(screen.getByRole('button', { name: 'LIST IS OPEN' }));
      expect(consoleSpy).toHaveBeenCalled();
    });

    it('calls the onOptionListClick callback when the OptionList main button is closed and is clicked', () => {
      const consoleSpy = jest
        .spyOn(console, 'error')
        .mockImplementation(() => {});

      render(
        <OptionList
          listContent={['foo', 'bar', 'bash']}
          isOpen={false}
          onOptionListClick={consoleSpy}
          onListItemClick={() => {}}
          openText="LIST IS OPEN"
          closedText="LIST IS CLOSED"
        />,
      );

      userEvent.click(screen.getByRole('button', { name: 'LIST IS CLOSED' }));
      expect(consoleSpy).toHaveBeenCalled();
    });

    it('does not call the onListItemClick callback if none of the list item buttons are clicked', () => {
      const consoleSpy = jest
        .spyOn(console, 'error')
        .mockImplementation(() => {});

      render(
        <OptionList
          listContent={['foo', 'bar', 'bash']}
          isOpen
          onOptionListClick={() => {}}
          onListItemClick={consoleSpy}
          openText="LIST IS OPEN"
          closedText="LIST IS CLOSED"
        />,
      );

      expect(consoleSpy).not.toHaveBeenCalled();
    });

    it('does not call the onListItemClick callback if none of the list item buttons are clicked but the main button is', () => {
      const consoleSpy = jest
        .spyOn(console, 'error')
        .mockImplementation(() => {});

      render(
        <OptionList
          listContent={['foo', 'bar', 'bash']}
          isOpen
          onOptionListClick={() => {}}
          onListItemClick={consoleSpy}
          openText="LIST IS OPEN"
          closedText="LIST IS CLOSED"
        />,
      );

      userEvent.click(screen.getByRole('button', { name: 'LIST IS OPEN' }));
      expect(consoleSpy).not.toHaveBeenCalled();
    });

    it('calls the onListItemClick callback once each time any of the list item buttons are clicked', () => {
      const consoleSpy = jest
        .spyOn(console, 'error')
        .mockImplementation(() => {});

      render(
        <OptionList
          listContent={['foo', 'bar', 'bash']}
          isOpen
          onOptionListClick={() => {}}
          onListItemClick={consoleSpy}
          openText="LIST IS OPEN"
          closedText="LIST IS CLOSED"
        />,
      );

      userEvent.click(screen.getByRole('button', { name: 'foo' }));
      userEvent.click(screen.getByRole('button', { name: 'bash' }));
      expect(consoleSpy).toHaveBeenCalledTimes(2);
    });

    it('calls the onListItemClick callback multiple times if a list item is clicked multiple times', () => {
      const consoleSpy = jest
        .spyOn(console, 'error')
        .mockImplementation(() => {});

      render(
        <OptionList
          listContent={['foo', 'bar', 'bash']}
          isOpen
          onOptionListClick={() => {}}
          onListItemClick={consoleSpy}
          openText="LIST IS OPEN"
          closedText="LIST IS CLOSED"
        />,
      );

      userEvent.click(screen.getByRole('button', { name: 'foo' }));
      userEvent.click(screen.getByRole('button', { name: 'foo' }));
      userEvent.click(screen.getByRole('button', { name: 'bash' }));
      expect(consoleSpy).toHaveBeenCalledTimes(3);
    });
  });

  describe('propTypes', () => {
    it('logs an error when attempting to render without the listContent prop', () => {
      const consoleSpy = jest
        .spyOn(console, 'error')
        .mockImplementation(() => {});

      expect(() => {
        render(
          <OptionList
            isOpen
            onOptionListClick={() => {}}
            onListItemClick={() => {}}
            openText="LIST IS OPEN"
            closedText="LIST IS CLOSED"
          />,
        );
      }).toThrow();
      expect(consoleSpy).toHaveBeenCalled();
    });

    it('logs an error when attempting to render the listContent prop as a non array', () => {
      const consoleSpy = jest
        .spyOn(console, 'error')
        .mockImplementation(() => {});

      expect(() => {
        render(
          <OptionList
            listContent="foo"
            isOpen
            onOptionListClick={() => {}}
            onListItemClick={() => {}}
            openText="LIST IS OPEN"
            closedText="LIST IS CLOSED"
          />,
        );
      }).toThrow();
      expect(consoleSpy).toHaveBeenCalled();
    });

    it('logs an error when attempting to render the listContent prop as an array of non strings', () => {
      const consoleSpy = jest
        .spyOn(console, 'error')
        .mockImplementation(() => {});

      render(
        <OptionList
          listContent={[true, false, true]}
          isOpen
          onOptionListClick={() => {}}
          onListItemClick={() => {}}
          openText="LIST IS OPEN"
          closedText="LIST IS CLOSED"
        />,
      );

      expect(consoleSpy).toHaveBeenCalled();
    });

    it('logs an error when attempting to render without the isOpen prop', () => {
      const consoleSpy = jest
        .spyOn(console, 'error')
        .mockImplementation(() => {});

      render(
        <OptionList
          listContent={['foo', 'bar', 'bash']}
          onOptionListClick={() => {}}
          onListItemClick={() => {}}
          openText="LIST IS OPEN"
          closedText="LIST IS CLOSED"
        />,
      );

      expect(consoleSpy).toHaveBeenCalled();
    });

    it('logs an error when attempting to render the isOpen prop as a non bool', () => {
      const consoleSpy = jest
        .spyOn(console, 'error')
        .mockImplementation(() => {});

      render(
        <OptionList
          listContent={['foo', 'bar', 'bash']}
          isOpen="baz"
          onOptionListClick={() => {}}
          onListItemClick={() => {}}
          openText="LIST IS OPEN"
          closedText="LIST IS CLOSED"
        />,
      );

      expect(consoleSpy).toHaveBeenCalled();
    });

    it('logs an error when attempting to render without the onOptionListClick prop', () => {
      const consoleSpy = jest
        .spyOn(console, 'error')
        .mockImplementation(() => {});

      render(
        <OptionList
          listContent={['foo', 'bar', 'bash']}
          isOpen
          onListItemClick={() => {}}
          openText="LIST IS OPEN"
          closedText="LIST IS CLOSED"
        />,
      );

      expect(consoleSpy).toHaveBeenCalled();
    });

    it('logs an error when attempting to render the onOptionListClick prop as a non function', () => {
      const consoleSpy = jest
        .spyOn(console, 'error')
        .mockImplementation(() => {});

      render(
        <OptionList
          listContent={['foo', 'bar', 'bash']}
          isOpen
          onOptionListClick={false}
          onListItemClick={() => {}}
          openText="LIST IS OPEN"
          closedText="LIST IS CLOSED"
        />,
      );

      expect(consoleSpy).toHaveBeenCalled();
    });

    it('logs an error when attempting to render without the onListItemClick prop', () => {
      const consoleSpy = jest
        .spyOn(console, 'error')
        .mockImplementation(() => {});

      render(
        <OptionList
          listContent={['foo', 'bar', 'bash']}
          isOpen
          onOptionListClick={() => {}}
          openText="LIST IS OPEN"
          closedText="LIST IS CLOSED"
        />,
      );

      expect(consoleSpy).toHaveBeenCalled();
    });

    it('logs an error when attempting to render the onListItemClick prop as a non function', () => {
      const consoleSpy = jest
        .spyOn(console, 'error')
        .mockImplementation(() => {});

      render(
        <OptionList
          listContent={['foo', 'bar', 'bash']}
          isOpen
          onOptionListClick={() => {}}
          onListItemClick={false}
          openText="LIST IS OPEN"
          closedText="LIST IS CLOSED"
        />,
      );

      expect(consoleSpy).toHaveBeenCalled();
    });

    it('logs an error when attempting to render without the openText prop', () => {
      const consoleSpy = jest
        .spyOn(console, 'error')
        .mockImplementation(() => {});

      render(
        <OptionList
          listContent={['foo', 'bar', 'bash']}
          isOpen
          onListItemClick={() => {}}
          onOptionListClick={() => {}}
          closedText="LIST IS CLOSED"
        />,
      );

      expect(consoleSpy).toHaveBeenCalled();
    });

    it('logs an error when attempting to render the openText prop as a non string', () => {
      const consoleSpy = jest
        .spyOn(console, 'error')
        .mockImplementation(() => {});

      render(
        <OptionList
          listContent={['foo', 'bar', 'bash']}
          isOpen
          onOptionListClick={() => {}}
          onListItemClick={() => {}}
          openText={false}
          closedText="LIST IS CLOSED"
        />,
      );

      expect(consoleSpy).toHaveBeenCalled();
    });

    it('logs an error when attempting to render without the closedText prop', () => {
      const consoleSpy = jest
        .spyOn(console, 'error')
        .mockImplementation(() => {});

      render(
        <OptionList
          listContent={['foo', 'bar', 'bash']}
          isOpen
          onListItemClick={() => {}}
          onOptionListClick={() => {}}
          openText="LIST IS OPEN"
        />,
      );

      expect(consoleSpy).toHaveBeenCalled();
    });

    it('logs an error when attempting to render the closedText prop as a non string', () => {
      const consoleSpy = jest
        .spyOn(console, 'error')
        .mockImplementation(() => {});

      render(
        <OptionList
          listContent={['foo', 'bar', 'bash']}
          isOpen
          onOptionListClick={() => {}}
          onListItemClick={() => {}}
          openText="LIST IS OPEN"
          closedText={false}
        />,
      );

      expect(consoleSpy).toHaveBeenCalled();
    });

    it('does not log any error when all of the needed props are provided', () => {
      const consoleSpy = jest
        .spyOn(console, 'error')
        .mockImplementation(() => {});

      render(
        <OptionList
          listContent={['foo', 'bar', 'bash']}
          isOpen
          onOptionListClick={() => {}}
          onListItemClick={() => {}}
          openText="LIST IS OPEN"
          closedText="LIST IS CLOSED"
        />,
      );

      expect(consoleSpy).not.toHaveBeenCalled();
    });
  });
});
