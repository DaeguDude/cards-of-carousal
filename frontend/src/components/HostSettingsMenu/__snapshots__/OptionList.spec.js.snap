// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OptionButton rendering matches the expected snapshot when active 1`] = `
.c0 {
  color: var(--primary-text-color);
  background-color: var(--primary-background-color);
  color: var(--secondary-text-color);
  background-color: var(--secondary-background-color);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  border: none;
  padding: 20px 25px;
  width: 100%;
  text-align: left;
  font-size: 1.5rem;
  font-weight: 700;
}

.c0:hover {
  background-color: var(--accent-background-color);
}

.c1 {
  color: var(--primary-text-color);
  background-color: var(--primary-background-color);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  border: none;
  padding: 20px 25px;
  width: 100%;
  text-align: left;
  font-size: 1.5rem;
  font-weight: 700;
}

.c1:hover {
  background-color: var(--accent-background-color);
}

.c2 {
  font-size: 1.3rem;
  padding: 10px 50px;
}

<div>
  <button
    className="c0"
    onClick={[Function]}
    type="button"
  >
    LIST IS ACTIVE
  </button>
  <button
    className="c1 c2"
    onClick={[Function]}
    type="button"
  >
    foo
  </button>
  <button
    className="c1 c2"
    onClick={[Function]}
    type="button"
  >
    bar
  </button>
  <button
    className="c1 c2"
    onClick={[Function]}
    type="button"
  >
    bash
  </button>
</div>
`;

exports[`OptionButton rendering matches the expected snapshot when not active 1`] = `
.c0 {
  color: var(--primary-text-color);
  background-color: var(--primary-background-color);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  border: none;
  padding: 20px 25px;
  width: 100%;
  text-align: left;
  font-size: 1.5rem;
  font-weight: 700;
}

.c0:hover {
  background-color: var(--accent-background-color);
}

<div>
  <button
    className="c0"
    onClick={[Function]}
    type="button"
  >
    LIST IS INACTIVE
  </button>
</div>
`;
